lib/libxdp/xsk.c

```diff
@@ -0,0 +1,982 @@
+	map_attr.value_size = sizeof(int);
+	map_attr.max_entries = 1;
+
+	map_fd = bpf_create_map_xattr(&map_attr);
+	if (map_fd < 0)
+		return detected;
+
+	insns[0].imm = map_fd;
+
+	memset(&prog_attr, 0, sizeof(prog_attr));
+	prog_attr.prog_type = BPF_PROG_TYPE_XDP;
+	prog_attr.insns = insns;
+	prog_attr.insns_cnt = ARRAY_SIZE(insns);
+	prog_attr.license = "GPL";
+
+	prog_fd = bpf_load_program_xattr(&prog_attr, NULL, 0);
+	if (prog_fd < 0) {
+		close(map_fd);
+		return detected;
+	}
+
+	ret = bpf_prog_test_run(prog_fd, 0, &data_in, 1, &data_out, &size_out, &retval, &duration);
+	if (!ret && retval == XDP_PASS)
+		detected = true;
+	close(prog_fd);
+	close(map_fd);
+	return detected;
+}
+
+static struct xdp_program *xsk_lookup_program(int ifindex)
+{
+	const char *version_name = "xsk_prog_version";
+	const char *prog_name = "xsk_def_prog";
+	struct xdp_multiprog *multi_prog;
+	struct xdp_program *prog = NULL;
+	__u32 version;
+	int err;
+
+	multi_prog = xdp_multiprog__get_from_ifindex(ifindex);
+	if (IS_ERR(multi_prog))
+		return NULL;
+
+	while ((prog = xdp_multiprog__next_prog(prog, multi_prog)))
+		if (!strcmp(xdp_program__name(prog), prog_name))
+			break;
+
+	if (!prog)
+		return NULL;
+
+	err = check_xdp_prog_version(xdp_program__btf(prog), version_name, &version);
+	if (err)
+		return ERR_PTR(err);
+	if (version > XSK_PROG_VERSION) {
+		pr_warn("XSK default program version %d higher than supported %d\n", version,
+			XSK_PROG_VERSION);
+		return ERR_PTR(-EOPNOTSUPP);
+	}
+
+	return prog;
+}
+
+static int __xsk_setup_xdp_prog(struct xsk_socket *xsk, int *xsks_map_fd)
+{
+	const char *fallback_prog = "xsk_def_xdp_prog_5.3.o";
+	const char *default_prog = "xsk_def_xdp_prog.o";
+	struct xsk_ctx *ctx = xsk->ctx;
+	struct xdp_program *prog;
+	const char *file_name;
+	int err;
+
+	prog = xsk_lookup_program(ctx->ifindex);
+	if (IS_ERR(prog))
+		return PTR_ERR(prog);
+
+	if (!prog) {
+		file_name = xsk_check_redirect_flags() ? default_prog : fallback_prog;
+		ctx->xdp_prog = xdp_program__find_file(file_name, NULL, NULL);
+		if (IS_ERR(ctx->xdp_prog))
+			return PTR_ERR(ctx->xdp_prog);
+
+		err = xsk_size_map(ctx->xdp_prog, ctx->ifname);
+		if (err)
+			goto err_prog_load;
+
+		err = xdp_program__attach(ctx->xdp_prog, ctx->ifindex,
+					  xsk_convert_xdp_flags(xsk->config.xdp_flags), 0);
+		if (err)
+			goto err_prog_load;
+
+	} else {
+		ctx->xdp_prog = xdp_program__clone(prog);
+		if (IS_ERR(ctx->xdp_prog))
+			return PTR_ERR(ctx->xdp_prog);
+	}
+
+	ctx->xsks_map_fd = xsk_lookup_bpf_map(xdp_program__fd(ctx->xdp_prog));
+	if (ctx->xsks_map_fd < 0) {
+		err = ctx->xsks_map_fd;
+		goto err_lookup;
+	}
+
+	if (xsk->rx) {
+		err = bpf_map_update_elem(ctx->xsks_map_fd, &ctx->queue_id, &xsk->fd, 0);
+		if (err)
+			goto err_lookup;
+	}
+	if (xsks_map_fd)
+		*xsks_map_fd = ctx->xsks_map_fd;
+
+	return 0;
+
+err_lookup:
+	if (!prog)
+		xdp_program__detach(ctx->xdp_prog, ctx->ifindex,
+				    xsk_convert_xdp_flags(xsk->config.xdp_flags), 0);
```

Comment:
This will detach the program even if it was cloned instead of attached above...

Reply:
Will fix.

Reply:
I believe it is correct as it is. The clone path is only taken if the local variable prog is set and prog is only set at the beginning before cloning or attaching. Note that all other calls use ctx->xdp_prog.